python_modules/libraries/dagster-airlift/dagster_airlift/in_airflow/base_asset_operator.py

```diff
@@ -214,22 +216,63 @@ def launch_runs_for_task(self, context: Context, dag_id: str, task_id: str) -> N
             context,
             session,
             dagster_url,
-            _build_dagster_run_execution_params(
+            build_dagster_run_execution_params(
                 tags,
                 job_identifier,
                 asset_key_paths=asset_key_paths,
             ),
         )
         logger.info("Waiting for dagster run completion...")
-        while status := self.get_dagster_run_status(session, dagster_url, run_id):
+        self.wait_for_run_and_retries(session=session, dagster_url=dagster_url, run_id=run_id)
+        logger.info("All runs completed successfully.")
+        return None
+
+    def wait_for_run(
+        self, session: requests.Session, dagster_url: str, run_id: str
+    ) -> DagsterRunResult:
+        while response := self.get_dagster_run_obj(session, dagster_url, run_id):
+            status = response["status"]
             if status in ["SUCCESS", "FAILURE", "CANCELED"]:
                 break
             time.sleep(self.dagster_run_status_poll_interval)
-        if status != "SUCCESS":
-            raise Exception(f"Dagster run {run_id} did not complete successfully.")
-        logger.info("All runs completed successfully.")
+        tags = {tag["key"]: tag["value"] for tag in response["tags"]}
+        return DagsterRunResult(status=response["status"], tags=tags)
+
+    def wait_for_run_and_retries(
+        self, session: requests.Session, dagster_url: str, run_id: str
+    ) -> None:
+        run_id_to_check = run_id
+        while True:
+            result = self.wait_for_run(
+                session=session, dagster_url=dagster_url, run_id=run_id_to_check
+            )
+            if result.success:
+                break
+            logger.info(f"Run {run_id_to_check} completed with status '{result.status}'.")
+            if result.run_will_automatically_retry and result.retried_run_id:
+                logger.info(
+                    f"Run {run_id_to_check} retried in run {result.retried_run_id}. Waiting for completion..."
+                )
+                run_id_to_check = result.retried_run_id
+                continue
+            elif result.run_will_automatically_retry:
+                logger.info(
+                    f"Run {run_id_to_check} failed, but is configured to automatically retry. Waiting for retried run to be created..."
+                )
+                continue
+            else:
+                raise Exception(f"Run {run_id_to_check} failed, and is not expected to retry.")
         return None
 
+    def make_runs_query_with_filter(
+        self, runs_filter: Mapping[str, Any], session: requests.Session, dagster_url: str
+    ) -> Sequence[Mapping[str, Any]]:
+        response = session.post(
+            f"{dagster_url}/graphql",
+            json={"query": RUNS_BY_TAG_QUERY, "variables": {"filter": runs_filter}},
+        )
+        return self.get_valid_graphql_response(response, "runsOrError")["results"]
```

Comment:
this is not used now and can be removed i think? If it is used somewhere, we should be a good example and add a limit

Reply:
I think you're right. Good call